/*
 * This file is part of the NUbots Codebase.
 *
 * The NUbots Codebase is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * The NUbots Codebase is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright 2013 NUBots <nubots@nubots.net>
 */

syntax = "proto3";

package messages.vision;

import "google/protobuf/timestamp.proto";
import "messages/MessageOptions.proto";
import "messages/Geometry.proto";
import "messages/Vector.proto";
import "messages/Matrix.proto";
import "messages/input/Sensors.proto";
import "messages/vision/ClassifiedImage.proto";

message VisionObject {
    message Measurement {
        vec3 position = 1;
        mat33 covariance = 2;
    }

    google.protobuf.Timestamp timestamp = 1;
    repeated Measurement measurement = 2;
    vec2 screen_angular = 3;
    vec2 angular_size = 4;

    input.Sensors sensors = 5        [(pointer)=SHARED];
    vision.ClassifiedImage image = 6 [(pointer)=SHARED];
}

message Ball {
    VisionObject vision_object = 1;
    Circle circle = 2;
}

message Goal {
    enum Side {
        UNKNOWN_SIDE = 0;
        LEFT = 1;
        RIGHT = 2;
    }

    enum Team {
        UNKNOWN_TEAM = 0;
        CYAN = 1;
        MAGENTA = 2;
    }

    VisionObject vision_object = 1;
    Side side = 2;
    Team team = 3;
    Quad quad = 4;
}

message Obstacle {
    VisionObject vision_object = 1;
    Polygon polygon = 2;
}
